File: Jump_to_Algorithm2/week32/SeoyeonHong/쿼드압축후개수세기.py
# https://school.programmers.co.kr/learn/courses/30/lessons/68936?language=python3

# 쿼드압축 후 0과 1의 개수 구하기
def solution(arr):
    answer = [0, 0] # 0과 1의 개수
    
    def compress(r, c, n): # 압축하고자 하는 영역의 시작 행, 열, 크기
        
        if n == 1: # 영역의 크기가 1이라면
            answer[arr[r][c]] += 1 # 해당 숫자 반환
            return

        quad = [] # 압축하고자 하는 숫자들의 배열
        for i in range(n):
            quad += arr[r+i][c:c+n]
            
        if len(set(quad)) == 1: # 모두 같은 수로 이루어져 있다면
            answer[quad[0]] += 1 # 0 또는 1의 개수 +1
        else: # 4개의 영역으로 나누어 압축
            n //= 2
            compress(r, c, n)
            compress(r, c+n, n)
            compress(r+n, c, n)
            compress(r+n, c+n, n)
        
        return
        
    compress(0, 0, len(arr[0])) # 배열 압축
    
    return answer

File: Jump_to_Algorithm2/week32/dayeongYoo/쿼드압축 후 개수 세기.java
public class Solution {
    int[] answer;

    public void quadZip(int[][] arr, int x, int y, int length) {
        // 현재 영역이 압축 가능하면 answer 배열에 값 추가하고 반환
        if (zipChk(arr, x, y, length, arr[x][y])) {
            if (arr[x][y] == 1) answer[1]++;
            else answer[0]++;
            return;
        }

        // 1번 영역
        quadZip(arr, x, y, length / 2);
        // 2번 영역
        quadZip(arr, x + length / 2, y, length / 2);
        // 3번 영역
        quadZip(arr, x, y + length / 2, length / 2);
        // 4번 영역
        quadZip(arr, x + length / 2, y + length / 2, length / 2);
    }

    public boolean zipChk(int[][] arr, int x, int y, int length, int arrVal) {
        for (int i = x; i < x + length; i++) {
            for (int j = y; j < y + length; j++) {
                // 영역의 첫번째 값과 나머지를 비교하다 다르면 반환
                if (arr[i][j] != arrVal) return false;
            }
        }
        return true;
    }

    public int[] solution(int[][] arr) {
        answer = new int[2];
        quadZip(arr, 0, 0, arr.length);
        return answer;
    }
}